bindings/java/src/main/com/apple/foundationdb/tuple/ByteArrayUtil.java

```diff
@@ -171,7 +173,69 @@ static boolean regionEquals(byte[] src, int start, byte[] pattern) {
 	 */
 	public static byte[] replace(byte[] src, int offset, int length,
 			byte[] pattern, byte[] replacement) {
-		return join(replacement, split(src, offset, length, pattern));
+		if(pattern == null || pattern.length == 0) {
+			return Arrays.copyOfRange(src, offset, offset + length);
+		}
+		ByteBuffer dest;
+		if(replacement == null || replacement.length != pattern.length) {
+			// Array might change size. This is the "tricky" case.
+			byte patternFirst = pattern[0];
+			int patternOccurrences = 0;
+			int currentPosition = offset;
+			while(currentPosition < offset + length) {
```

Comment:
It could be an over kill but if this will be called frequently enough, try implementing the [KMP search](https://en.wikipedia.org/wiki/Knuth–Morris–Pratt_algorithm) could be much faster than this. KMP algorithm does take extra space, so if this is memory sensitive then this is good enough.

Reply:
I have no idea how often this method is called. We don't actually invoke it anywhere in our code base (not counting tests or for the purpose of providing optional parameters), so it is only invoked by consumers of our API.

In practice, the other variant of `replace` (which takes a `ByteBuffer` destination) is always invoked by our own code using `\x00` as the pattern and `\x00\xff` as the replacement or vice versa. I don't think KMP would help us that much in that case, but I could be wrong.

Reply:
Yes if the search term is short, the difference won't be large (say `n` is the length of the size of the search to search in and `k` is the size of the term, it's  `O(n*k)` vs `O(n+k)`)